src/components/common/TypeaheadSelect/TypeaheadSelect.tsx

```diff
@@ -1,443 +1,201 @@
-  const [filterValue, setFilterValue] = useState<string>('');
-  const [isFiltering, setIsFiltering] = useState<boolean>(false);
-  const [focusedItemIndex, setFocusedItemIndex] = useState<number | null>(null);
-  const [activeItemId, setActiveItemId] = useState<string | null>(null);
-  const textInputRef = useRef<HTMLInputElement>();
+  const [isFiltering, setIsFiltering] = useState(false);
+  const [inputValue, setInputValue] = useState('');
+  const inputRef = useRef<HTMLInputElement>(null);
 
-  const NO_RESULTS = t('No results');
+  useImperativeHandle(ref, () => inputRef.current!, []);
 
-  const selected = useMemo(
-    () => selectOptions.find((option) => option.value === props.selected || option.isSelected),
-    [props.selected, selectOptions],
+  const selectedOption = useMemo(
+    () => options.find((option) => option.value === value),
+    [options, value],
   );
 
-  const filteredSelections = useMemo(() => {
-    let newSelectOptions: TypeaheadSelectOption[] = selectOptions;
-
-    // Filter menu items based on the text input value when one exists
-    if (isFiltering && filterValue) {
-      newSelectOptions = filterFunction(filterValue, selectOptions);
-
-      if (
-        isCreatable &&
-        filterValue.trim() &&
-        !newSelectOptions.find(
-          (option) => String(option.content).toLowerCase() === filterValue.toLowerCase(),
-        )
-      ) {
-        const createOption = {
-          content:
-            typeof createOptionMessage === 'string'
-              ? createOptionMessage
-              : createOptionMessage(filterValue),
-          value: filterValue,
-        };
-        newSelectOptions = isCreateOptionOnTop
-          ? [createOption, ...newSelectOptions]
-          : [...newSelectOptions, createOption];
-      }
-
-      // When no options are found after filtering, display 'No results found'
-      if (!newSelectOptions.length) {
-        newSelectOptions = [
-          {
-            content:
-              typeof noOptionsFoundMessage === 'string'
-                ? noOptionsFoundMessage
-                : noOptionsFoundMessage(filterValue),
-            isAriaDisabled: true,
-            value: NO_RESULTS,
-          },
-        ];
-      }
-    }
+  const filteredOptions = useMemo(
+    () =>
+      generateFilteredOptions({
+        createOptionMessage,
+        filterFunction,
+        inputValue,
+        isCreatable,
+        isFiltering,
+        noResultsMessage,
+        options,
+      }),
+    [
+      isFiltering,
+      inputValue,
+      options,
+      filterFunction,
+      isCreatable,
+      createOptionMessage,
+      noResultsMessage,
+    ],
+  );
 
-    // When no options are  available,  display 'No options available'
-    if (!newSelectOptions.length) {
-      newSelectOptions = [
+  // Show filtered options when filtering, all options when not, or no options message
+  const displayOptions = useMemo(() => {
+    if (options.length === 0) {
```

Comment:
Please use isEmpty

Reply:
Added new warning rules for usage of isEmpty because this is something I frequently forget to use, and will help us catch more of these without reviewers having to focus so much on it.